game-app/game-headed/src/main/java/games/strategy/engine/framework/startup/ui/ServerSetupPanel.java

```diff
@@ -334,44 +336,84 @@ private void internalPlayerListChanged() {
-      final PlayerRow newPlayerRow =
-          new PlayerRow(
-              entry.getKey(),
-              reloadSelections,
-              entry.getValue(),
-              new PlayerTypes(HeadedPlayerTypes.getPlayerTypes()));
-      playerRows.add(newPlayerRow);
-      newPlayerRow.update(players, playersEnabled);
+      final String playerName = entry.getKey();
+      GamePlayer player = null;
+      if (gameSelectorModel.getGameData() != null) {
+        player = gameSelectorModel.getGameData().getPlayerList().getPlayerId(playerName);
+      }
+      PlayerTypes.Type playerType =
+          determinePlayerType(playerName, player, reloadSelections, visiblePlayerTypes);
+      PlayerRow row = new PlayerRow(playerName, playerType, visiblePlayerTypes, entry.getValue());
+      row.update(players, playersEnabled);
+      playerRows.add(row);
     }
     layoutPlayers();
     internalPlayersTakenChanged();
   }
 
+  private PlayerTypes.Type determinePlayerType(
+      String playerName,
+      @Nullable GamePlayer player,
+      Map<String, String> reloadSelections,
+      List<PlayerTypes.Type> visiblePlayerTypes) {
+    final String previousSelection = reloadSelections.get(playerName);
+    if (previousSelection != null && previousSelection.equalsIgnoreCase("Client")) {
+      return HeadedPlayerTypes.HUMAN_PLAYER;
+    } else if (previousSelection != null && !previousSelection.equals("no_one")) {
+      Optional<PlayerTypes.Type> type =
+          visiblePlayerTypes.stream().filter(t -> t.getLabel().equals(previousSelection)).findAny();
+      if (type.isPresent()) {
+        return type.get();
+      }
+    }
+
+    if (player != null && player.isDefaultTypeAi()) {
+      return PlayerTypes.PRO_AI;
+    } else if (player != null && player.isDefaultTypeDoesNothing()) {
+      return HeadedPlayerTypes.DOES_NOTHING_PLAYER;
+    } else if (playerName.startsWith("Neutral") || playerName.startsWith("AI")) {
+      return PlayerTypes.PRO_AI;
+    } else {
+      return HeadedPlayerTypes.HUMAN_PLAYER;
+    }
+  }
+
   class PlayerRow {
     private final JLabel nameLabel;
     private final JLabel playerLabel;
     private final JCheckBox localCheckBox;
     private final JCheckBox enabledCheckBox;
-    private final JComboBox<String> type;
+    private final JComboBox<PlayerTypes.Type> type;
     private final JButton alliance;
-    private final ActionListener localPlayerActionListener =
-        new ActionListener() {
-          @Override
-          public void actionPerformed(final ActionEvent e) {
+
+    PlayerRow(
+        final String playerName,
+        final PlayerTypes.Type playerType,
+        final Collection<PlayerTypes.Type> playerTypes,
+        final Collection<String> playerAlliances) {
+      nameLabel = new JLabel(playerName);
+      playerLabel = new JLabel(model.getMessenger().getLocalNode().getName());
+      type = new JComboBox<>(new DefaultComboBoxModel<>(new Vector<>(playerTypes)));
```

Comment:
IIRC we had a helper method somewhere that takes a list and constructs a combo box model out of it, but no idea where or if we still have it

Reply:
Found it, thanks!